control_plane/attachment_service/src/service.rs

```diff
@@ -551,6 +583,69 @@ impl Service {
             }
         }
     }
+    #[instrument(skip_all)]
+    async fn spawn_heartbeat_driver(&self) {
+        self.startup_complete.clone().wait().await;
+
+        const HEARTBEAT_INTERVAL: Duration = Duration::from_secs(5);
+
+        let mut interval = tokio::time::interval(HEARTBEAT_INTERVAL);
+        while !self.cancel.is_cancelled() {
+            tokio::select! {
+              _ = interval.tick() => { }
+              _ = self.cancel.cancelled() => return
+            };
+
+            let nodes = {
+                let locked = self.inner.read().unwrap();
+                locked.nodes.clone()
+            };
+
+            let res = self.heartbeater_handler.heartbeat(nodes).await;
+            if let Ok(deltas) = res {
+                let mut locked = self.inner.write().unwrap();
+                let (nodes, tenants, scheduler) = locked.parts_mut();
+                let mut new_nodes = (**nodes).clone();
+
+                tracing::info!("new_nodes={new_nodes:?}");
+                for (node_id, state) in deltas.0 {
+                    tracing::info!("HB: {node_id} -> {state:?}");
+                    let node = match new_nodes.get_mut(&node_id) {
+                        Some(node) => node,
+                        None => continue,
+                    };
+
+                    match state {
+                        PageserverState::Available { utilization, .. } => {
+                            node.set_availability(NodeAvailability::Active(UtilizationScore(
+                                utilization.utilization_score,
+                            )));
+                        }
+                        PageserverState::Offline => {
+                            node.set_availability(NodeAvailability::Offline);
+                        }
+                    }
+
+                    scheduler.node_upsert(node);
+                }
+
+                *nodes = Arc::new(new_nodes);
+
+                for (tenant_shard_id, tenant_state) in tenants {
+                    if tenant_state.should_schedule(nodes) {
```

Comment:
Moved this to the background reconciliation loop in https://github.com/neondatabase/neon/pull/7092/commits/088fbd3acf1f18952908b92f6ef00193c73fb08d.

>  The logic of should_schedule can be simpler: it just needs to check if the attached location is on an unavailable node, and if it is, we would call demote_attached, then schedule -- that way we keep secondaries around across periods of unavailability.

I'm not convinced about this. We can also have situations where there's no intent attached location. We should probably catch those as well. Also, I don't see why we'd have to call `demote_attached` here. It's already done in `node_configure` for transitions to offline.

Reply:
Let's take this part into a separate PR -- we can do merge heartbeats first and then refine the background scheduling part.

>We can also have situations where there's no intent attached location

Yes, true -- what I mean is that those cases are already covered by schedule().  So I'm thinking we would have a step that does the demote_attached for any unavailable attached nodes, and then calls schedule() (no need for a maybe_ variant because schedule is already a no-op if constraints are already satisfied)